sdk/eventhub/azure-eventhubs/azure/eventhub/_producer_client.py

```diff
@@ -203,7 +190,9 @@ def close(self):
                 :caption: Close down the client.
 
         """
-        for p in self._producers:
-            if p:
-                p.close()
+        with self._lock:
+            for producer in self._producers.values():
+                if producer:
```

Comment:
How producers list could contain something that evaluate to False?

Reply:
Because when we get the partition IDs, we don't upfront create a producer for each ID, so the value can be None:
https://github.com/yunhaoling/azure-sdk-for-python/blob/feature/eventhubs_refactor/sdk/eventhub/azure-eventhubs/azure/eventhub/_producer_client.py#L77

The producer itself isn't created until a message needs to be sent to that partition.